### PR TITLE
Memory block allocator

### PR DESCRIPTION
**Pull request by TheJJ**
This is mainly a splitup and fixup of @schets #150. Opinions?

### CODE REVIEW

File: cpp/util/block_allocator.h [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+	std::swap(this->lookup_vec[this->last_good], this->lookup_vec[ind]);
+	this->lookup_vec[this->last_good]->ptr_ind = this->last_good;
+	this->lookup_vec[ind]->ptr_ind = ind;
+}
+
+template<class T>
+template<bool do_free>
+inline void block_allocator<T>::releaser(T *to_release) {
+	if (likely(to_release)) {
+		val_ptr *vptr = (val_ptr*)to_release;
```
Comment (fr3akout) [inline cpp/util/block_allocator.h]: please use c++ style casts

File: cpp/util/block_allocator.h [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+	}
+	blkp[this->block_size-1].next = nullptr;
+	this->lookup_vec[0]->num_alloc = 0;
+	this->lookup_vec[0]->first_open = &blkp[0];
+}
+
+template<class T>
+block_allocator<T>::_block::_block(size_t data_len, size_t ind)
+	:
+	data((val_ptr*)malloc(data_len*sizeof(val_ptr)), internal_deleter()),
```
Comment (fr3akout) [inline cpp/util/block_allocator.h]: please use c++ style casts

File: cpp/util/block_allocator.h [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+	//!Resets the allocator to starting state, does not call destructors
+	void deallocate();
+};
+
+
+template<class T>
+fixed_block_allocator<T>::fixed_block_allocator(size_t _block_size)
+	:
+	block_size(_block_size),
+	data((val_ptr*)malloc(block_size*sizeof(val_ptr)), internal_deleter()),
```
Comment (fr3akout) [inline cpp/util/block_allocator.h]: please use c++ style casts

File: cpp/util/block_allocator.h [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+		return (T*)fst;
+	}
+	return nullptr;
+}
+
+template<class T>
+template<bool do_free>
+void fixed_block_allocator<T>::releaser(T *to_release) {
+	if (likely(to_release)) {
+		val_ptr *vptr = (val_ptr*)to_release;
```
Comment (fr3akout) [inline cpp/util/block_allocator.h]: please use c++ style casts

File: cpp/util/stack_allocator.h [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+template<class T>
+stack_allocator<T>::stack_allocator(size_t _stack_size)
+	:
+	stack_size(_stack_size) {
+	this->add_substack();
+}
+
+template<class T>
+void stack_allocator<T>::add_substack() {
+	T *space = (T*)malloc(this->stack_size * sizeof(T));
```
Comment (fr3akout) [inline cpp/util/stack_allocator.h]: please use c++ style casts

File: cpp/util/stack_allocator.h [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+	 * Resets the allocator to starting state, may destruct
+	 */
+	template<bool destruct>
+	void doclear();
+};
+
+template<class T>
+fixed_stack_allocator<T>::fixed_stack_allocator(size_t stack_size)
+	:
+	data((T*)malloc(stack_size*sizeof(T)), deleter()),
```
Comment (fr3akout) [inline cpp/util/stack_allocator.h]: please use c++ style casts

**Comment by TheJJ** [comment]
now depends on #232 btw.

**Comment by franciscod** [comment]
looks good! tests, yay!

**Comment by fr3akout** [comment]
I don't really see the need for a custom allocator at the current stage of openage. I would only implement or merge one, if there are really some performance issues due to memory management. Otherwise one simply adds a lot of code, that can be pretty error-prone and doesn't bring any noticeable improvements.

**Comment by franciscod** [comment]
i think it would bring unneeded complexity at this point... maybe we should wait until problems appear and then look again at this one?

**Comment by TheJJ** [comment]
we'll revive the code once we need the block alloc again.

**Pull request closed without merging**